cmd/kaniko-acr/main.go

```diff
@@ -0,0 +1,388 @@
+		// only setup auth when pushing or credentials are defined
+		if registry == "" {
+			return nil, fmt.Errorf("registry must be specified")
+		}
+
+		token, err := getAcrToken(tenantId, clientId, clientSecret, certPath, registry)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to fetch acrToken")
+		}
+		fmt.Printf("acr token " + token)
+		err = createDockerCfgFile(username, token, registry)
+	}
+
+	return dockerConfig, nil
+}
+
+func getAcrToken(tenantId, clientId, clientSecret, certPath, registry string) (string, error) {
+	if tenantId == "" {
+		return "", fmt.Errorf("tenantId can't be empty foe AAD authentication")
+	}
+
+	if clientId == "" {
+		return "", fmt.Errorf("clientId can't be empty foe AAD authentication")
+	}
+
+	if clientSecret == "" && certPath == "" {
+		return "", fmt.Errorf("one of accessKey or secretKey should be defined")
+	}
+
+	// TODO check for presence of file as well.
+	os.Setenv(accessKeyEnv, clientId)
+	os.Setenv(secretKeyEnv, clientSecret)
+	os.Setenv(tenantKeyEnv, tenantId)
+	env, err := azidentity.NewEnvironmentCredential(nil)
+	context.Background()
+	policy := policy.TokenRequestOptions{
+		Scopes: []string{"https://management.azure.com/.default"},
+	}
+
+	os.Unsetenv(accessKeyEnv)
+	os.Unsetenv(secretKeyEnv)
+	os.Unsetenv(tenantKeyEnv)
+
+	azToken, err := env.GetToken(context.Background(), policy)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch access token")
+	}
+
+	acrToken, err := fetchAcrToken(tenantId, azToken.Token, registry)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
+	}
+	return acrToken, nil
+}
+
+func fetchAcrToken(tenantId, token, registry string) (string, error) {
+	formData := url.Values{
+		"grant_type":   {"access_token"},
+		"service":      {registry},
+		"tenant":       {tenantId},
+		"access_token": {token},
+	}
+	jsonResponse, err := http.PostForm(fmt.Sprintf("https://%s/oauth2/exchange", registry), formData)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch acr token")
+	}
+	var response map[string]interface{}
+	json.NewDecoder(jsonResponse.Body).Decode(&response)
+	return response["refresh_token"].(string), nil
+}
+
+// Create the docker config file for authentication
+func createDockerCfgFile(username, password, registry string) error {
+	if username == "" {
+		return fmt.Errorf("Username must be specified")
+	}
+	if password == "" {
+		return fmt.Errorf("Password must be specified")
+	}
+
+	err := os.MkdirAll(dockerPath, 0600)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to create %s directory", dockerPath))
+	}
+
+	authBytes := []byte(fmt.Sprintf("%s:%s", username, password))
+	encodedString := base64.StdEncoding.EncodeToString(authBytes)
+	jsonBytes := []byte(fmt.Sprintf(`{"auths": {"%s": {"auth": "%s"}}}`, "https://"+registry, encodedString))
+	fmt.Print("-- " + username + " -- " + string(jsonBytes))
+	err = ioutil.WriteFile(dockerConfigPath, jsonBytes, 0644)
```

Comment:
We are writing to the same file at L249 as well

Reply:
removed